File: src/main/java/komu/blunt/eval/VariableInfo.java
package komu.blunt.eval;

import komu.blunt.objects.Symbol;

import static com.google.common.base.Preconditions.checkNotNull;

public final class VariableInfo {

    public final Symbol name;
    public final int offset;

    public VariableInfo(Symbol name, int offset) {
        if (offset < 0) throw new IllegalArgumentException("negative offset: " + offset);

        this.name = checkNotNull(name);
        this.offset = offset;
    }

    @Override
    public String toString() {
        return name.toString();
    }
}

File: src/main/kotlin/komu/blunt/analyzer/VariableInfo.kt
package komu.blunt.analyzer

import komu.blunt.objects.Symbol

data class VariableInfo(val name: Symbol, val offset: Int) {

    init {
        require(offset >= 0) { "negative offset $offset" }
    }

    fun toReference(frame: Int) = VariableReference.nested(frame, offset, name)

    override fun toString() = name.toString()
}
